### PR TITLE
Fix insertion of complex JSONs with nested arrays [2]

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Addition to #36077.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Now fuzzer from #35962 passes. Need to improve it to find more bugs.

File: src/DataTypes/ObjectUtils.cpp [hunk 1]
```diff
@@ -189,9 +203,15 @@ void checkObjectHasNoAmbiguosPaths(const PathsInData & paths)
...
-                throw Exception(ErrorCodes::DUPLICATE_COLUMN,
+                throw Exception(ErrorCodes::INCOMPATIBLE_COLUMNS,
                     "Data in Object has ambiguous paths: '{}' and '{}'",
                     paths[i].getPath(), paths[j].getPath());
+
+            if (hasDifferentStructureInPrefix(paths[i].getParts(), paths[j].getParts()))
+                throw Exception(ErrorCodes::INCOMPATIBLE_COLUMNS,
+                    "Data in Object has ambiguous paths: '{}' and '{}'. "
+                    "Paths have prefixes matched by names, but different in structure",
+                    paths[i].getPath(), paths[j].getPath());
```
Comment (CurtizJ) [inline src/DataTypes/ObjectUtils.cpp:214]: Fix for `01825_type_json_14.sql`.

File: src/DataTypes/Serializations/JSONDataParser.cpp [hunk 2]
```diff
@@ -213,11 +213,14 @@ bool JSONDataParser<ParserImpl>::tryInsertDefaultFromNested(
 {
     /// If there is a collected size of current Nested
     /// then insert array of this size as a default value.
+    if (path.empty() || array.empty())
+        return false;
 
-    if (path.empty())
+    /// Last element is not Null, because otherwise this path wouldn't exist.
+    auto nested_key = getNameOfNested(path, array.back());
+    if (nested_key.empty())
```
Comment (CurtizJ) [inline src/DataTypes/Serializations/JSONDataParser.cpp:221]: Fix for `01825_type_json_15.sh`

File: src/DataTypes/Serializations/JSONDataParser.cpp [hunk 3]
```diff
@@ -253,7 +256,18 @@ StringRef JSONDataParser<ParserImpl>::getNameOfNested(const PathInData::Parts &
...
+    /// several arrays with the same prefix, but with independent sizes.
+    /// Consider we have array element with type `k2 Tuple(k3 Nested(...), k5 Nested(...))`
+    /// Then subcolumns `k2.k3` and `k2.k5` may have indepented sizes and we should extract
+    /// `k3` and `k5` keys instead of `k2`.
+
+    for (size_t i = 0; i < path.size(); ++i)
+        if (path[i].is_nested)
+            return StringRef{path[i].key};
+
+    return {};
```
Comment (CurtizJ) [inline src/DataTypes/Serializations/JSONDataParser.cpp:270]: Fix for `01825_type_json_16.sh`

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
@CurtizJ look at ClickHouse special build check failure

**Pull request merged**